File: authentication/php/phpleague/src/Oauth/OnBehalfOfContext.php
<?php
/**
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.
 * Licensed under the MIT License.  See License in the project root
 * for license information.
 */


namespace Microsoft\Kiota\Authentication\Oauth;

/**
 * Class OnBehalfOfContext
 *
 * Token request parameters for the on_behalf_of flow using a secret
 *
 * @package Microsoft\Kiota\Authentication
 * @copyright 2022 Microsoft Corporation
 * @license https://opensource.org/licenses/MIT MIT License
 * @link https://developer.microsoft.com/graph
 */
class OnBehalfOfContext extends BaseSecretContext
{
    /**
     * @var string
     */
    private string $assertion;
    /**
     * @var array<string, string>
     */
    private array $additionalParams;

    /**
     * @param string $tenantId
     * @param string $clientId
     * @param string $clientSecret
     * @param string $assertion initial access token sent to the current API
     * @param array<string, string> $additionalParams extra params to be added to token request
     */
    public function __construct(string $tenantId, string $clientId, string $clientSecret, string $assertion, array $additionalParams = [])
    {
        if (!$assertion) {
            throw new \InvalidArgumentException("Assertion cannot be empty");
        }
        $this->assertion = $assertion;
        $this->additionalParams = $additionalParams;
        parent::__construct($tenantId, $clientId, $clientSecret);
    }

    /**
     * @inheritDoc
     */
    public function getParams(): array
    {
        return array_merge($this->additionalParams, parent::getParams(), [
            'assertion' => $this->assertion,
            'grant_type' => $this->getGrantType(),
            'requested_token_use' => 'on_behalf_of'
        ]);
    }

    /**
     * @inheritDoc
     */
    public function getGrantType(): string
    {
        return 'urn:ietf:params:Oauth:grant-type:jwt-bearer';
    }
}

File: authentication/ruby/contexts/on_behalf_of_context.rb
# frozen_string_literal: true

module MicrosoftKiotaAbstractions
    # Token request context class for the on behlaf of grant type.
    class OnBehalfOfContext
        attr_reader :grant_type, :additional_params, :tenant_id, :client_id, :client_secret

        # This is the initializer for OnBehalfOfContext, the token request context when
        # using the client credential grant flow. 
        # :params
        #   tenant_id: a string containing the tenant id 
        #   client_id: a string containing the client id 
        #   client_secret: a string containing the client secret 
        #   assertion: string containing assertion (access token used in the request)
        #   additional_params: hash of symbols to string values, ie { response_mode: 'fragment', prompt: 'login' }
        #                      default is empty hash
        def initialize(tenant_id, client_id, client_secret, assertion, additional_params = {})
          raise StandardError, 'assertion cannot be empty' if assertion.nil? || assertion.empty?

          @tenant_id = tenant_id
          @client_id = client_id
          @client_secret = client_secret
          @assertion = assertion
          @additional_params = additional_params
          @grant_type = 'urn:ietf:params:Oauth:grant-type:jwt-bearer'
    
          if tenant_id.nil? || client_id.nil? || client_secret.nil? || client_secret.empty? || tenant_id.empty? || client_id.empty?
            raise StandardError, 'tenant_id, client_secret, and client_id cannot be empty'
          end

        end
    end
end
